### PR TITLE
(feature/0457-support-amazon-cloudfront-function): Support Amazon Clo…

### PR DESCRIPTION
**Pull request by jolo-dev**
Collect Amazon CloudFront function #457 

## Solution

Add to Cloudfront Functions a new file `functions.go`

## Changes Made

- new file `provider/aws/cloudfront/functions.go`

## How to Test

Create a new Cloudfront functions: https://us-east-1.console.aws.amazon.com/cloudfront/v3/home?region=us-east-1#/functions/create

## Screenshots

Screenshot 2023-10-22 at 14 45 32

## Notes

[Any additional notes or information that you would like to share with the reviewers.]

## Checklist

- [X] Code follows the <a href="https://github.com/tailwarden/komiser/blob/master/CONTRIBUTING.md">contributing</a> guidelines
- [X] Changes have been thoroughly tested
- [X] <a href="https://github.com/tailwarden/docs.komiser.io">Documentation</a> has been updated, if necessary
- [X] Any dependencies have been added to the project, if necessary

### CODE REVIEW

File: providers/aws/cloudfront/functions.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+func Functions(ctx context.Context, client ProviderClient) ([]Resource, error) {
+	resources := make([]Resource, 0)
+	var config cloudfront.ListFunctionsInput
+	cloudfrontClient := cloudfront.NewFromConfig(*client.AWSClient)
+
+	tempRegion := client.AWSClient.Region
+	client.AWSClient.Region = "us-east-1"
+	cloudwatchClient := cloudwatch.NewFromConfig(*client.AWSClient)
+	client.AWSClient.Region = tempRegion
```
Comment (Traxmaxx) [inline providers/aws/cloudfront/functions.go:31]: 🙌

**Comment by jolo-dev** [comment]
@Traxmaxx the dashboard fails because of the missing Toast component: https://discord.com/channels/932683789384183808/933353577102061669/1165250794417836072

**Comment by Traxmaxx** [comment]
> @Traxmaxx the dashboard fails because of the missing Toast component: https://discord.com/channels/932683789384183808/933353577102061669/1165250794417836072

Hey thanks! I'm aware and try to get someone from the maintainers to help me review open PRs to get the fix in asap!

**Review by mlabouardy — APPROVED** [review]

**Comment by bishal7679** [comment]
@mlabouardy This changes has already been raised by this PR #1131
and had to close it due to checks failing!

**Comment by AvineshTripathi** [comment]
Hey @bishal7679 your PR also adds something some details to distributions, I would recommend you to review this PR to see if it matches to what your changes are and we can remove those from your PR. However we will make sure you get added as a collaborator for this PR.

This is just to make sure both get credits for the amazing work, however i can only do that when both you and @jolo-dev are ready for it. Thanks!

**Comment by bishal7679** [comment]
> Hey @bishal7679 your PR also adds something some details to distributions, I would recommend you to review this PR to see if it matches to what your changes are and we can remove those from your PR. However we will make sure you get added as a collaborator for this PR.
> 
> This is just to make sure both get credits for the amazing work, however i can only do that when both you and @jolo-dev are ready for it. Thanks!

@AvineshTripathi I completely support this! Wouldn't it be good?? get PR #1131 merged first [PS:- might get complicated for merge conflict] after then this PR #1127 could be addon since both are adding same file `functions.go` and in #1131 dynamic pricing and more metrics being used and #1127 its invocation metric can be added there to make the calculation more efficient??

**Comment by jolo-dev** [comment]
@AvineshTripathi I believe the one of #1131  is a bit more sophisticated. However, there might be a few changes needed. He got the Duration which I couldn't find. @bishal7679 Have you tried and that worked?
L138 and L158.
And I am not sure about the Namespace of L92.

**Comment by AvineshTripathi** [comment]
Great! I keep this to you'll and let me know how you want to proceed or need any assitance.

**Review by bishal7679 — CHANGES_REQUESTED** [review]
Hope it helps!
REF :- [metricname](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/programming-cloudwatch-metrics.html#cloudfront-metrics-function-values)

File: providers/aws/cloudfront/functions.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+			invocations := 0.0
+			if metricsInvocationsOutput != nil && len(metricsInvocationsOutput.Datapoints) > 0 {
+				invocations = *metricsInvocationsOutput.Datapoints[0].Sum
+			}
+			if invocations > freeTierInvocations {
+				invocations -= freeTierInvocations
+			}
+
+			monthlyCost := invocations * 0.0000001
```
Comment (Azanul) [inline providers/aws/cloudfront/functions.go:70]:
```suggestion
			monthlyCost := invocations * costPerInvocation
```

File: providers/aws/cloudfront/functions.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/aws/aws-sdk-go-v2/service/cloudfront"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatch"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatch/types"
+	. "github.com/tailwarden/komiser/models"
+	. "github.com/tailwarden/komiser/providers"
+	"github.com/tailwarden/komiser/utils"
+)
+
+const (
+	freeTierInvocations = 2000000
```
Comment (Azanul) [inline providers/aws/cloudfront/functions.go:20]:
```suggestion
	freeTierInvocations = 2000000
	costPerInvocation = 0.0000001
```

**Review by Azanul — CHANGES_REQUESTED** [review]

**Review by Azanul — APPROVED** [review]

**Review by bishal7679 — APPROVED** [review]
LGTM!

**Pull request merged**